src/test/java/rcms/utilities/daqexpert/reasoning/logic/basic/CloudFUnumberTest.java

```diff
@@ -0,0 +1,146 @@
+package rcms.utilities.daqexpert.reasoning.logic.basic;
+
+import java.io.File;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+import org.junit.Test;
+import static org.junit.Assert.*;
+import rcms.utilities.daqaggregator.data.DAQ;
+import rcms.utilities.daqaggregator.persistence.StructureSerializer;
+import rcms.utilities.daqexpert.Setting;
+
+/**
+ *
+ * @author holzner
+ */
+public class CloudFUnumberTest
+{
+
+	private CloudFUnumber makeInstance() {
+		CloudFUnumber result = new CloudFUnumber();
+
+		// mock properties
+		Properties properties = new Properties();
+		properties.setProperty(Setting.EXPERT_LOGIC_CLOUDFUNUMBER_THRESHOLD_TOTAL_FRACTION.getKey(), "0.03");
+		properties.setProperty(Setting.EXPERT_LOGIC_CLOUDFUNUMBER_HOLDOFF_SECONDS.getKey(), "1800");
+
+		result.parametrize(properties);
+
+		return result;
+	}
+
+	/**
+	 * method to load and deserialize a snapshot given a file name
+	 *
+	 * similar to FlowchartCaseTestBase.getSnapshot() but with
+	 * a different base directory.
+	 */
+	public static DAQ getSnapshot(String fname) throws URISyntaxException {
+
+		StructureSerializer serializer = new StructureSerializer();
+
+		URL url = Deadtime.class.getResource(fname);
+
+		File file = new File(url.toURI());
+
+		return serializer.deserialize(file.getAbsolutePath());
+	}
+
+	/** @param expectedResult is the result which should be returned
+	 *    by the module, taking into account the holdoff time
+   *
+   * @param expectedResultBeforeHoldOff is the expected 'raw' result
+	 *    before applying the holdoff timer
+	 */
+	private void doChecks(CloudFUnumber instance, String snapshotFname,
+					boolean expectedResult,
+					boolean expectedResultBeforeHoldOff,
+					String expectedMessage) throws URISyntaxException {
+
+		DAQ daq = getSnapshot(snapshotFname);
+
+		Map<String, Boolean> results = new HashMap<String, Boolean>();
+
+		boolean result = instance.satisfied(daq, results);
+		assertEquals("result after taking into account holdoff for snapshot " + snapshotFname,
+						expectedResult, result);
+
+		// check the 'raw' result before applying the holdoff timer
+		boolean resultBeforeHoldOff = ! instance.cloudCanBeOn(daq);  //instance.getHoldOffTimer().getInput();
```

Comment:
this is not entirely what you test. In tests you say 'module should fire if there was no holdoff'. The method cloudCanBeOn does not represent the whole condition of the LM. The condition of this LM is 'is there more Fus in cloude mode than x%' not 'can cloud be one right now' which is only a part of condition. Thus I cannot use this mechanism to other LM (let's say Deadtime LM) I would have to implement part of condition for all of them. What I ment by generic mechanism is: give me satisfy method result with or without the holdoff. The condition stay exactly the same (it's encapsulated in satisfy method which doesn't include a single line of holdoff calculation). Here I have to remember to make 2 calls to holdoff object. I propose something similar to this:

https://github.com/cmsdaq/DAQExpert/commit/50d7701aae96dd82c3e79042d92c635cb9a037e9

Reply:
as quickly discussed yesterday, holdoff in `CloudFuNumberTest` is applied in a slightly different way than `HoldOffLm` implements:

- `HoldOffLm` only returns true when the 'raw' condition has been continuously true for the holdoff period
- `CloudFuNumberTest` only returns true if the LHC beam mode has been one of the non-cloud ones for at least the holdoff period but then fires immediately if some FUs are still found in cloud mode.

If `CloudFuNumberTest` were to inherit from `HoldOffLm` its `satisfied()` method would return `LHC in non-cloud mode && FUs in cloud mode found` and then apply a holdoff on this. In case the cloud is accidentally started during stable beams it would still wait for the holdoff period to raise an alert.

Reply:
I now renamed variables in the new class `HoldOffTimer` of this pull request to make them consistent with the ones used in `HoldOffLm` and also changed the logic accordingly (to calculate the expiry timestamp once instead of calculating the elapsed time since start) (commits a1f9c02, 370a10e, b251116 and 5f6ff4c).